sycl/source/detail/usm/usm_dispatch.cpp

```diff
@@ -0,0 +1,346 @@
+  pfn_##_funcname = (_funcname##_fn)clGetExtensionFunctionAddressForPlatform(  \
+      platform, #_funcname);
+
+USMDispatcher::USMDispatcher(cl_platform_id platform) {
+  // TODO: update when platform_impl becomes more PI aware
+  
+  if (pi::useBackend(pi::Backend::SYCL_BE_PI_OPENCL)) {
+    GET_EXTENSION(clHostMemAllocINTEL);
+    GET_EXTENSION(clDeviceMemAllocINTEL);
+    GET_EXTENSION(clSharedMemAllocINTEL);
+    GET_EXTENSION(clMemFreeINTEL);
+    GET_EXTENSION(clGetMemAllocInfoINTEL);
+    GET_EXTENSION(clSetKernelArgMemPointerINTEL);
+    GET_EXTENSION(clEnqueueMemsetINTEL);
+    GET_EXTENSION(clEnqueueMemcpyINTEL);
+    GET_EXTENSION(clEnqueueMigrateMemINTEL);
+    GET_EXTENSION(clEnqueueMemAdviseINTEL);
+    mEmulated = !(pfn_clHostMemAllocINTEL && pfn_clDeviceMemAllocINTEL &&
+                  pfn_clSharedMemAllocINTEL && pfn_clMemFreeINTEL &&
+                  pfn_clSetKernelArgMemPointerINTEL &&
+                  pfn_clEnqueueMemsetINTEL && pfn_clEnqueueMemcpyINTEL);
+    mEmulator.reset(new CLUSM());
+  }
+  // Else Error?
+}
+
+void *USMDispatcher::hostMemAlloc(pi_context Context,
+                                  cl_mem_properties_intel *Properties,
+                                  size_t Size, pi_uint32 Alignment,
+                                  pi_result *ErrcodeRet) {
+  void *RetVal = nullptr;
+
+  if (pi::useBackend(pi::Backend::SYCL_BE_PI_OPENCL)) {
+    cl_context CLContext = pi::cast<cl_context>(Context);
+
+    if (mEmulated) {
+      RetVal = mEmulator->hostMemAlloc(CLContext, Properties, Size, Alignment,
+                                       pi::cast<cl_int *>(ErrcodeRet));
+    } else {
+      RetVal = pfn_clHostMemAllocINTEL(CLContext, Properties, Size, Alignment,
+                                       pi::cast<cl_int *>(ErrcodeRet));
+    }
+  }
+
+  if (ErrcodeRet && !RetVal) {
+    ErrcodeRet[0] = PI_INVALID_OPERATION;
+  }
+  return RetVal;
+}
+
+void *USMDispatcher::deviceMemAlloc(pi_context Context, pi_device Device,
+                                    cl_mem_properties_intel *Properties,
+                                    size_t Size, pi_uint32 Alignment,
+                                    pi_result *ErrcodeRet) {
+  void *RetVal = nullptr;
+
+  if (pi::useBackend(pi::Backend::SYCL_BE_PI_OPENCL)) {
+    cl_context CLContext = pi::cast<cl_context>(Context);
+    cl_device_id CLDevice = pi::cast<cl_device_id>(Device);
+
+    if (mEmulated) {
+      RetVal = mEmulator->deviceMemAlloc(CLContext, CLDevice, Properties, Size,
+                                         Alignment,
+                                         pi::cast<cl_int *>(ErrcodeRet));
+    } else {
+      RetVal = pfn_clDeviceMemAllocINTEL(CLContext, CLDevice, Properties, Size,
+                                         Alignment,
+                                         pi::cast<cl_int *>(ErrcodeRet));
+    }
+  }
+
+  if (ErrcodeRet && !RetVal) {
+    ErrcodeRet[0] = PI_INVALID_OPERATION;
```

Comment:
Shouldn't it be:
*ErrcodeRet = RetVal;

?

Reply:
Sure.

Reply:
The point is changing from PI_INVALID_OPERATION to RetVal.

Reply:
No - it shouldn't.  This path should be triggered if we have non-null ErrcodeRet (aka one was provided) and RetVal remains nullptr.  RetVal here isn't an error code - it's a void* ptr from allocation.

If you prefer, I could change this to "else if (ErrcodeRet && !RetVal)" as anything under the first if should have an error code already.

Reply:
No - misread the initial comment.

Reply:
- the !Retval in that case